images/build/debian-iptables/distroless/Dockerfile

```diff
@@ -0,0 +1,63 @@
+# Copyright 2022 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARG GORUNNERIMAGE
+# Using directly debian base, as we are going to just copy the required binaries
+FROM debian:bullseye as build
+
+COPY stage-binary-and-deps.sh /
+COPY get-binaries-from-package.sh /
+
+ARG STAGE_DIR="/opt/stage"
+RUN apt -y update && \
+    apt -y install bash curl && \
+    mkdir -p "${STAGE_DIR}" && \
+    /get-binaries-from-package.sh conntrack "${STAGE_DIR}" && \
+    /get-binaries-from-package.sh ebtables "${STAGE_DIR}" && \
+    /get-binaries-from-package.sh ipset "${STAGE_DIR}" && \
+    /get-binaries-from-package.sh iptables "${STAGE_DIR}" && \
+    /get-binaries-from-package.sh kmod "${STAGE_DIR}" && \
+    `# dash is the real Debian basic shell` \
+    /stage-binary-and-deps.sh /bin/dash "${STAGE_DIR}" && \
+    `# below binaries are used by iptables-wrapper-installer.sh` \
+    /stage-binary-and-deps.sh /bin/cat "${STAGE_DIR}" && \
+    /stage-binary-and-deps.sh /bin/chmod "${STAGE_DIR}" && \
+    /stage-binary-and-deps.sh /bin/grep "${STAGE_DIR}" && \
+    /stage-binary-and-deps.sh /bin/ln "${STAGE_DIR}" && \
+    /stage-binary-and-deps.sh /bin/rm "${STAGE_DIR}" && \
+    /stage-binary-and-deps.sh /usr/bin/wc "${STAGE_DIR}" && \
+    `# iptables modules needs to be copied` \
+    cp -r /usr/lib/"$(uname -m)"-linux-gnu/xtables/ "${STAGE_DIR}"/usr/lib/"$(uname -m)"-linux-gnu/ && \
+    curl -o "${STAGE_DIR}"/iptables-wrapper-installer.sh https://raw.githubusercontent.com/kubernetes-sigs/iptables-wrappers/master/iptables-wrapper-installer.sh && \
+    chmod +x "${STAGE_DIR}"/iptables-wrapper-installer.sh && \
+    ln -sf /bin/dash "${STAGE_DIR}"/bin/sh
+
+# we're going to borrow the /go-runner binary in the final image
+# dedupe this binary by just copying it from the go-runner image
+FROM ${GORUNNERIMAGE} as gorunner
+
+# We need to use distroless:base here as tzdata, glibc and some other base packages
+# are required
+FROM gcr.io/distroless/base as intermediate
+# FROM debian:bullseye
+COPY clean-distroless.sh /clean-distroless.sh
+COPY --from=build /opt/stage /
+COPY --from=gorunner /go-runner /go-runner
+# iptables-wrapper-installer needs to know that iptables exists before doing all its magic
+RUN ln -sf /usr/sbin/iptables-legacy /usr/sbin/iptables && \
+    /iptables-wrapper-installer.sh && \
+    /clean-distroless.sh
+
+FROM scratch
+COPY --from=intermediate / /
```

Comment:
This "flattens" the image but does not look logic if you don't know that. Please add a comment unless this is a well-known pattern. Um, please add a comment anyway.

Reply:
#### A general comment on bandwidth

This PR is mainly about reducing bandwidth for download of `kube-proxy`. While the distroless flattened image is smaller than the current kube-proxy image it *may* not reduce bandwidth if the "debian layer" is used by other images in the system. I assume that downloading works like `docker pull` in the way that layers already downloaded are not re-downloaded.

Using this function: 

```sh
cmd_docker_ls_layers() {
	test -n "$1" || die "No image"
	mkdir -p $tmp
	docker inspect "$1" > $tmp/out || die "Failed; docker inspect"
	docker save $1 > $tmp/img.tar || die "Failed; docker save"

	local l
	for l in $(tar tf $tmp/img.tar | grep layer.tar); do
		echo "=== Layer; $l"
		tar -xOf $tmp/img.tar $l | tar t
	done
}
```
to check the layers in `k8s.gcr.io/kube-proxy:v1.23.5`:
```
=== Layer; 0d4c84ad80a8ef647af040e64a332fa13efa51f8f976c067c5b69d893c320988/layer.tar
go-runner
=== Layer; 75a6d8043f1f07d1b960d20f210f70a87f544eee34a79952c8ab89ba876042df/layer.tar
usr/
usr/local/
usr/local/bin/
usr/local/bin/build.log
usr/local/bin/kube-proxy
=== Layer; ebe8a644e5b8e1739a4863db4fa67c22807d8486425a85b77a1a4a89a235c94b/layer.tar
.dockerenv
bin/
bin/cat
bin/chgrp
...  (the rest is Debian)
```
This mean that only the first 2 layers are downloaded if the "debian layer" is already on the system.

Now, the "debian layer" must be *exactly* the same, which may be rare, like "never happens", but if the same debian image used in another commonly used K8s image? Then the bandwidth problem may just shift to that image since it must now download the "debian image" that it previously got from the `kube-proxy` download.

Reply:
@uablrek on worker nodes other k8s image probably are not there?? it's just kubelet and kube-proxy

Reply:
I think the ones in https://github.com/kubernetes/kubernetes/issues/109406#issuecomment-1094408522 could be checked

Reply:
But in any case, I think a distroless image is a good thing.